feature/various_fixes_and_tweaks

nindim opened the pull request:
- Added AutoplayType, this allows animations to trigger in either Awake or OnEnable (defaults to Awake to retain backwards compatibility)
- Fixed issues with backward playback in Editor Preview Tool and at runtime.
- Fixed callbacks not firing correctly when playing backward.
- If AutoplayType is OnEnable, then OnDisable will kill the tweens and reset the animated objects to their initial states, this ensures that they are in the right state for the Tweens to be regenerated the next time the animation is enabled.
- TimeScale slider in Preview Tool can now be used before playback commences.
- Exposed getters for PlayOnAwake and PauseOnAwake, these can be used for asserts to catch incorrect setup of AniamtionSequenceControllers from the code's perspective.
- Fixed an issue with "Move To Anchored Position" when using the IsRelative flag. It must be set when calling tween.From(isRelative) for it to work.
- Fixed delays not working for Callback steps.
- Fixed potential null refs in various DOTweenActions.
- Allowed various parts of AnimationSequenceController and AnimationSequenceControllerCustomEditor to be overridden.
- Play On Awake and Pause On Awake now alter their inspector label dependent on the selected AutoplayType.
- Fixed a bug with the completion callbacks passed to Play, PlayForward and PlayBackwards, old callbacks were executing for subsequent plays. All listeners are now removed each time any playback function is executed.
- Fixed typo in PlayForward bool.

nindim committed:
- Added AutoplayMode, this allows animations to trigger in either Awake or OnEnable (defaults to Awake to retain backwards compatibility)
- Fixed issues with backward playback in Editor Preview Tool and at runtime.
- Fixed callbacks not firing correctly when playing backward.
- If AutoplayMode is OnEnable, then OnDisable will kill the tweens and reset the animated objects to their initial states, this ensures that they are in the right state for the Tweens to be regenerated the next time the animation is enabled.
- TimeScale slider in Preview Tool can now be used before playback commences.
- Exposed getters for PlayOnAwake and PauseOnAwake, these can be used for asserts to catch incorrect setup of AniamtionSequenceControllers from the code's perspective.
- Fixed an issue with "Move To Anchored Position" when using the IsRelative flag. It must be set when calling tween.From(isRelative) for it to work.
- Fixed delays not working for Callback steps.
- Fixed potential null refs in various DOTweenActions.
- Allowed various parts of AnimationSequenceController and AnimationSequenceControllerCustomEditor to be overridden.

brunomikoski commented on Scripts/Runtime/Core/AnimationSequencerController.cs:
```diff
@@ -22,11 +28,16 @@ public enum PlayType
         [SerializeField]
-        private bool playOnAwake;
+        private AutoplayMode autoplayMode = AutoplayMode.Awake;
+        public AutoplayMode AutoPlayMode { get { return autoplayMode;} set { autoplayMode = value; } }
```
Can you create a `SetAutoPlayMode();` method instead of a setter?

nindim replied:
Hey Bruno, I will remove the setter for now without adding the SetAutoPlayMode() function as it is not actually used.

nindim replied:
Entire property has been removed.

brunomikoski commented on Scripts/Runtime/Core/AnimationSequencerController.cs:
```diff
@@ -36,21 +47,38 @@ public enum PlayType
         [SerializeField]
         private UnityEvent onStartEvent = new UnityEvent();
-        public UnityEvent OnStartEvent => onStartEvent;
+        public UnityEvent OnStartEvent { get { return onStartEvent;} protected set {onStartEvent = value;}}
```
In this case I believe would be better to just add a `AddOnStartEventCallbackListener(UnityAction callback)`; just to make sure someone cannot just erase all the callbacks by creating a new unity event using a setter by mistake.

nindim replied:
Hey Bruno, the setter is protected and I am using this in my derived class to redirect OnStartEvent and OnFinsihedEvent based on the playback direction. It feels to me like the AddOnStartEventCallbackListener() etc would be different public functionality?

brunomikoski commented on Scripts/Runtime/Core/AnimationSequencerController.cs:
```diff
@@ -36,21 +47,38 @@ public enum PlayType
         [SerializeField]
         private UnityEvent onFinishedEvent = new UnityEvent();
-        public UnityEvent OnFinishedEvent => onFinishedEvent;
+        public UnityEvent OnFinishedEvent { get { return onFinishedEvent;} protected set {onFinishedEvent = value;}}
```
same as https://github.com/brunomikoski/Animation-Sequencer/pull/35/files#r767199752

nindim replied:
Same as: https://github.com/brunomikoski/Animation-Sequencer/pull/35#discussion_r767258983

nindim committed:
- Play On Awake and Pause On Awake now alter their inspector label dependent on the selected AutoplayType.
- Renamed AutoplayMode to AutoplayType.
- Fixed a bug with the completion callbacks passed to Play, PlayForward and PlayBackwards, old callbacks were executing for subsequent plays. All listeners are now removed each time any playback function is executed.
- Removed public property for AutoplayMode as it's not required.
- Fixed typo in PlayForward bool.
